### PR TITLE
Added shared drive crawler

### PR DESCRIPTION
**Pull request by Pourliver**
I added the automatic shared drive crawler I had in mind, as shared a little in #137.

This has been done after #140, so its been easier to implement since I knew the download flow and quirks.

I had to create a crawler_config folder since I didn't like the idea of mixing classes definition and plaintext files.

I tried to put more documentation in this class since the core logic may be hard to follow. I've tested it a lot and it works amazingly well, but please do test it.

### CODE REVIEW

**Comment by Res260** [comment]
Idea: log in a file each file, even if not downloaded

File: pyrdp/mitm/FileCrawler.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        localPath = f"{basePath}{remotePath}"
+
+        self.log.info("Saving %(remotePath)s to %(localPath)s", {"remotePath": remotePath, "localPath": localPath})
+
+        try:
+            # Create parent directory, don't raise error if it already exists
+            Path(localPath).parent.mkdir(parents=True, exist_ok=True)
+            targetFile = open(localPath, "wb")
+        except Exception as e:
+            self.log.error("Cannot save file: %(exception)s", {"exception": str(e)})
```
Comment (Res260) [inline pyrdp/mitm/FileCrawler.py]: `log.exception` would be better

File: pyrdp/mitm/FileCrawler.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+            return
+
+        self.downloadFiles[remotePath] = targetFile
+        self.deviceRedirection.sendForgedFileRead(file.deviceID, remotePath)
+
+    def listDirectory(self, deviceID: int, path: str, download: bool = False):
```
Comment (Res260) [inline pyrdp/mitm/FileCrawler.py]: docstring would help

File: pyrdp/mitm/config.py [hunk 3]
```diff
@@ -53,6 +53,15 @@ def __init__(self):
         self.payloadDuration: int = None
         """Amount of time the payload should take to complete, in milliseconds"""
 
+        self.disableCrawler: bool = False
+        """Whether the crawler should be crawling or not"""
```
Comment (Res260) [inline pyrdp/mitm/config.py]: should be disabled or not*, to match the boolean value

File: pyrdp/mitm/crawler_config/match.txt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+*/*.csv
+*/*.xls*
+*/*.doc*
+*/*.ppt*
+
+# Possible malware samples
+*/*.bat
+*/*.exe
+*/*.dll
+*/*.ps1
```
Comment (Res260) [inline pyrdp/mitm/crawler_config/match.txt]: .py? possibly other languages as well

File: bin/pyrdp-mitm.py [hunk 5]
```diff
@@ -164,6 +164,9 @@ def main():
     parser.add_argument("--payload-powershell-file", help="PowerShell script to run automatically upon connection (as -EncodedCommand)", default=None)
     parser.add_argument("--payload-delay", help="Time to wait after a new connection before sending the payload, in milliseconds", default=None)
     parser.add_argument("--payload-duration", help="Amount of time for which input / output should be dropped, in milliseconds. This can be used to hide the payload screen.", default=None)
+    parser.add_argument("--disable-crawler", help="Disable automatic shared drive scraping", action="store_true")
+    parser.add_argument("--crawler-match-file", help="File to be used by the crawler to chose what to download when scraping the client shared drives.")
+    parser.add_argument("--crawler-ignore-file", help="File to be used by the crawler to chose what folders to avoid when scraping the client shared drives.")
```
Comment (Res260) [inline bin/pyrdp-mitm.py]: put default value?

File: pyrdp/mitm/crawler_config/match.txt (hunk 4 above)
Comment (Pourliver) [reply pyrdp/mitm/crawler_config/match.txt]: Do you have any input on this? @obilodeau This file is what we currently download from the disk, folder AND files.

**Comment by Pourliver** [comment]
> Idea: log in a file each file, even if not downloaded

To add to this idea, since we talked about it in person, the goal is to keep track of what the crawler is seeing, so we can further improve the default match / ignore regexes later.

File: pyrdp/mitm/crawler_config/match.txt (hunk 4 above)
Comment (Pourliver) [reply pyrdp/mitm/crawler_config/match.txt]: I think that crawling code files would give us a lot of thrash and would expand the crawl time. I went with compiled files and scripts that work out-of-the-box on Windows / Linux.

File: bin/pyrdp-mitm.py (hunk 5 above)
Comment (Pourliver) [reply bin/pyrdp-mitm.py]: Since the configuration takes places in `bin/pyrdp-mitm.py` and the default file is in `pyrdp/mitm/crawler_config/match.txt`, I don't think moving the default value in the configuration parser is a good idea. Don't forget that the `bin/` content gets moved to the user binaries folder during installation, and we lose direct path references.

File: pyrdp/mitm/FileCrawler.py [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+        :param filePath: Unix-style path of the file.
+        :param isDirectory: True if the file is a directory.
+        """
+
+        self.deviceID = deviceID
+        self.name = name
+        self.path = filePath
+        self.isDirectory = isDirectory
+
+class FileCrawler(DeviceRedirectionMITMObserver):
```
[the remark this answers is not in the document]
Comment (Pourliver) [reply pyrdp/mitm/FileCrawler.py]: I'm not too sure the value that it brings, but I can rename it

File: bin/pyrdp-mitm.py (hunk 5 above)
Comment (Res260) [reply bin/pyrdp-mitm.py]: makes sense

File: pyrdp/mitm/FileCrawler.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Res260) [reply pyrdp/mitm/FileCrawler.py]: It was for consistence and verboseness, but I dont think we're constant in the rest of the code base so doesnt matter for now

**Comment by Pourliver** [comment]
The PR is ready for another review. I added a new logger that is not from the root logger, so we could log in another file without messing with STDOUT. This new file keeps track of every file we crawled.

**Comment by Res260** [comment]
Oops commented on the commet instead of the PR:

Besides the exception() log TODO, LGTM!

**Comment by Pourliver** [comment]
Sorry, but I'm not too sure I understand whats wrong @Res260 , could you please elaborate a bit?

**Comment by Res260** [comment]
I think I saw you added a log.exception() somewhere else in the code, so I
thought it would be easy to implement instead of leaving a TODO in the code.

**Comment by Pourliver** [comment]
You're right, I totally missed this one, my bad! Fixed in latest commit.

**Comment by Res260** [comment]
Lgtm

**Comment by Pourliver** [comment]
Fixed a small problem with the parser where the last line of each file got truncated, and simplified the match file since `*/*.doc*` can also match `*/*.docker`, which is unintended.

File: bin/pyrdp-mitm.py [hunk 7]
```diff
@@ -169,6 +191,9 @@ def main():
     parser.add_argument("--payload-powershell-file", help="PowerShell script to run automatically upon connection (as -EncodedCommand)", default=None)
     parser.add_argument("--payload-delay", help="Time to wait after a new connection before sending the payload, in milliseconds", default=None)
     parser.add_argument("--payload-duration", help="Amount of time for which input / output should be dropped, in milliseconds. This can be used to hide the payload screen.", default=None)
+    parser.add_argument("--disable-crawler", help="Disable automatic shared drive scraping", action="store_true")
```
Comment (xshill) [inline bin/pyrdp-mitm.py]: I think this is such a big feature that it should be enabled consciously, so I'll change this to opt-in.
